californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java

```diff
@@ -1256,4 +1256,75 @@ public void testObserveFailureBeforeToSendDuringBlockNotification() throws Excep
 
 		// @after check there is no leak
 	}
+
+	/**
+	 * Ensure that a notification is not handled as a proactive cancel response
+	 * 
+	 * <pre>
+	 * (actual used MIDs and Tokens my vary!)
+	 * (Established observe relation)
+	 * CON [MID=55453, T=[15d4ebce227d141d]], GET, /test, observe(0)    ----->
+	 * <-----   ACK [MID=55453, T=[15d4ebce227d141d]], 2.05, observe(1)
+	 * <-----   CON [MID=8001, T=[15d4ebce227d141d]], 2.05, observe(3)
+	 * ACK [MID=8001]   ----->
+	 * CON [MID=55454, T=[15d4ebce227d141d]], GET, /test, observe(1)    ----->
+	 * <-----   CON [MID=8002, T=[15d4ebce227d141d]], 2.05, observe(4)
+	 * RST [MID=8002]   ----->
+	 * <-----   NON [MID=8003, T=[15d4ebce227d141d]], 2.05, observe(5)
+	 * RST [MID=8003]   ----->
+	 * CON [MID=55454, T=[15d4ebce227d141d]], GET, /test, observe(1)    ----->
+	 * CON [MID=55454, T=[15d4ebce227d141d]], GET, /test, observe(1)    ----->
+	 * <-----   CON [MID=8004, T=[15d4ebce227d141d]], 2.05
+	 * ACK [MID=8004]   ----->
+	 * </pre>
+	 */
+	@Test
+	public void testNotificationIsNotHandledAsProactiveCancelResponse() throws Exception {
```

Comment:
I would prefer to have different tests for:
- receiving notifications during the cancel request.
- receiving notifications after successful cancel request 
- receiving notifications after failed cancel request

Reply:
This test is only about : 
 - receiving notifications during the cancel request.

(I changed it a little bit after our discussion in #688, before to push this PR)

Would you like I add tests for : 
- receiving notifications after successful cancel request
- receiving notifications after failed cancel request

Currently the behavior for failed and successful cancel will be the same.

Reply:
> Currently the behavior for failed and successful cancel will be the same.

Do you mean, currently 
- receiving notifications after successful cancel request
- receiving notifications after failed cancel request

behaves equal? 
I would guess (hope :-) ), the client will RST notifications, if the cancel is successful.

Reply:
> I would guess (hope :-) ), the client will RST notifications, if the cancel is successful.

After cancelling, notifications are not RST anymore. (success or not)
That's what I try to explain in https://github.com/eclipse/californium/issues/688#issuecomment-403501405.

There is 2 API, coapclient and coapendpoint.
Coapclient is more high level api than coapendpoint.

With coapclient, you can use CoapObserveRelation.proactiveCancel() and it will behave like you expect.

But coapendpoint, is low-level api, there is one API to send proactive cancel request and one to remove observation at client side (passive cancellation).

FMPOV in a perfect world, with coapendpoint API notification should be always handle as long as we don't remove observation from obervation store.
But there is a side effect and during an ongoing proactive cancel request, notifications are RST. 
As you explain fixing this means adding corner case in the stack and maybe this is not really worthy. In generally, the proactive use case is to do both at the same time. I mean removing observation from the store, then sending a cancel request.

With the low level API, the success of cancel request does not determine if the observation is removed or not. This should be done manually before or after using endpoint.cancelObservation()

In this test, I just send the cancel request.

Reply:
Sorry Simon,
your tests seems for me to be smoke tests, where messages are fired in a chaotic, none-compliant way. I thought, your intention is to ensure, that there are no leaks introduced. If I have over seen the variant, where the cancel is responded successful and further notifications are accepted instead of rejected, then this is my fault. But even in the comment you refer, the cancel request is NOT responded with success, instead it timesout, which I would count as "failed"

That's also my comment:
Please provide not only the "smoke/chaos" test, provide also the three compliant test-cases.

Reply:
After removing "the smoke", I recognized, that this test doesn't use the `CoapClient`. Therefore the "proactive cancel" doesn't work. I prepared a fix for proactive cancel without `CoapClient`, but using "observe" without a `CoapClient` will not filter the notifications based on their order and requires a `NotificationListener`. 
So at least, this seems for me to be not recommended.
I got also aware, that `Request.waitForResponse()` is still in between. The comments about the notifications are not cleaned up and the implementation sticks also on the old notification processing.

Reply:
> your intention is to ensure, that there are no leaks introduced.

My intention was to test the use case you described at https://github.com/eclipse/californium/issues/688#issuecomment-402757040

> But even in the comment you refer, the cancel request is NOT responded with success,

I would like to fix this comment but I don't know which one you are talking about. Could you precise which one ?

> Please provide not only the "smoke/chaos" test, provide also the three compliant test-cases.

I think this is not so chaotic, just 2 notifications which comes before we receive the proactive cancel and a lost ACK for the cancel separated response. (This is compliant case)

but I could add more tests, but currently I'm not sure what is the behavior expected.

> After removing "the smoke"

I still don't understand what you consider as smoke in this test. :-/

> but using "observe" without a CoapClient will not filter the notifications based on their order and requires a NotificationListener.
So at least, this seems for me to be not recommended.

I think this is the interest of having this low-level API. Bringing more flexibility.

> I got also aware, that Request.waitForResponse() is still in between. The comments about the notifications are not cleaned up and the implementation sticks also on the old notification processing.

I do not understand the issue here with the use of waitForResponse, I do not want to get notification, I want to ensure that notification received is not consider as response for the proactive cancel request.

Reply:
"chaotic" means, it's not that easy/obvious, what it tests. At least for me ...

Reply:
> "chaotic" means, it's not that easy/obvious, what it tests. At least for me ...

I do not take your remarks badly, I just don't know how to do this better. :-)
How could I test that a notification is not handled as a proactive cancel response in a more easy/obvious to understand way ?

(Maybe just remove the lost ACK ?)

Reply:
> I would like to fix this comment but I don't know which one you are talking about. Could you precise which one ?

```
X------ ACK M3              ------ // proactive ACK lost
<------ CON M6, T1          ------ // response of proactive cancel
```

That's what is "chaotic" for me. First "ACK lost", and then response. So the test-case doesn't simply piggybacked response to the request. That makes it hard to see, if it fails, because the test is broken or because the implementation is broken. To understand such stuff, much more time is required.

Reply:
OK I changed this.

Reply:
> OK I changed this.

Any due day? For that issue, we may also merge it (after https://github.com/eclipse/californium/pull/692#discussion_r201326176) and add this stuff later.

Reply:
I would like to finished some tasks, I postponed too many time for Leshan. I hope I could rework on this before the end of the day.

Did you ask this because you would like to release the M11 ? If yes, I think the tests in this PR is not mandatory for M11.

Reply:
Mentally, I shifted M11 already for a week :-).